File: 049/cpp/main.cpp
class Solution {
public:
    string gen(string s) {
        map<char, int> m;
        for (auto &c: s) {
            m[c] += 1;
        }
        string str;
        for (auto &c: "abcdefghijklmnopqrstuvwxyz") {
            str += to_string(m[c]);
            str += ":";
        }
        return str;
    }

    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        map<string, vector<string>> groups;
        for (auto &str: strs) {
            auto key = gen(str);
            groups[key].push_back(str);
        }
        vector<vector<string>> res;
        for (auto &[key, val]: groups) {
            res.push_back(val);
        }
        return res;
    }
};

File: 049/rust/main.rs
use std::collections::HashMap;

impl Solution {
    fn generate(str: &String) -> String {
        let mut m = HashMap::new();
        let str = str.as_bytes();
        for b in str {
            let t = m.entry(b).or_insert(0);
            *t += 1
        }
        let mut s = String::new();
        for b in "abcdefghijklmnopqrstuvwxyz".as_bytes() {
            let t = m.entry(b).or_insert(0);
            s += &(*t).to_string();
            s += ":"
        }
        return s;
    }

    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut groups = HashMap::new();
        for str in strs {
            let key = Solution::generate(&str);
            let t = groups.entry(key).or_insert(Vec::new());
            (*t).push(str.clone());
        }
        let mut res = Vec::new();
        for (k, v) in groups {
            res.push(v);
        }
        return res;
    }
}
